### PR TITLE
feat(localUsage): read Claude Code token spend from its own transcripts

### PR DESCRIPTION
**Pull request by murdore**
## What this is

First slice of the observability lane — the half of the CLI-support plan that was entirely unbuilt.

The proxy's ledger can only account for traffic that went through it, which caps coverage at the CLIs exposing a base-URL override. Vendors control that, and most don't. Every CLI writes a local transcript regardless, so reading those covers the rest — and recovers history from before the proxy was ever installed. It's also an **independent source of truth**: a pricing defect in the proxy's own accounting can't hide from a reader that derives cost from a different input.

Scoped to **one reader**. The plan specified twelve; building them together means the whole thing lands untested or not at all.

## Two findings from real data that changed the design

**Top-level session transcripts are a rounding error.** `~/.claude/projects` on this machine holds **17,439 transcripts totalling 9.7 GB**, and only ~100 are `<project>/<sessionId>.jsonl`. The rest live under `<sessionId>/subagents/`. A reader globbing only `<project>/*.jsonl` reports a fraction of a subagent-heavy user's spend while looking perfectly healthy. This one recurses.

**9.7 GB is also why `scan()` is time-scoped.** `sinceDays` defaults to 30 and filters on mtime before opening anything; `Infinity` is there for a deliberate full sweep. Measured on real data:

```
sinceDays=1   4,520ms   395 files    44,426 turns   $11,989
sinceDays=7   6,794ms   2,003 files  121,918 turns  $21,751
```

## Design notes worth reviewing

**Dedup keeps the largest output count, not the first or last.** A resumed session re-logs turns it already wrote, and the re-logged copy can carry a higher *or* lower output count. Summing double-counts; first- and last-write-wins are each wrong half the time. The map is per-file and discarded after, so it stays bounded across thousands of files.

**Cost confidence stays `"modeled"` when some turns go unpriced.** `"heuristic"` is reserved for a figure that was itself estimated. Here every dollar came from a real rate table — what's missing is *turns*, not accuracy, and that's reported exactly by `unpricedRequests` / `unpricedModels` (on this machine: Claude Code's internal `<synthetic>`, plus a stray local `qwen2.5:0.5b`). Downgrading the row would make a genuinely estimated total indistinguishable from this one.

**`readAllLocalUsage` is exported from the package entry deliberately.** Without a public export the only way to test this would be importing out of `src/lib/` — which CLAUDE.md rule 15 names as disallowed. The export is what makes the feature testable at all, not an afterthought.

## Proof

Eight cases, all driving `../dist/index.js` — the same entry an SDK consumer gets. Six against a fixture HOME, two against this machine's real transcripts (they SKIP when `~/.claude/projects` is absent).

Every invariant confirmed by breaking it — **separately**, because breaking two at once proved less than it looked like: killing the directory walk made every fixture read zero rows and masked the dedup result entirely.

```
dedup -> last-write-wins      ✗ expected 340, got 280
no recursion into subagents   ✗ expected 2 requests, got 1
fixed                         ✓ 8 passed
```

**One thing worth flagging: the dedup case did not catch its own regression the first time.** Its fixture had the duplicate pair only ever ascending — the ordering where keep-max and last-write-wins agree — so it passed cleanly against a deliberately broken reader. It now carries a descending duplicate too, so first-write-wins and last-write-wins each fail it. Same false-green shape as the Codex configurator round-trip found earlier in this stack; a round-trip test that has never been shown to fail is not coverage.

## Gates

```
tsc --noEmit                  0 errors
pnpm run lint                 0 errors (56 pre-existing warnings)
test:local-usage              8 passed, 0 failed
```

## Not in this slice

Eleven other readers; the persisted incremental cursor (the mtime filter covers the immediate performance problem); cross-file dedup for resumed sessions that span files; any CLI subcommand. `LocalUsageCliId` already names all twelve CLIs so the next reader is an additive registration, not a refactor.

## Summary by CodeRabbit

- **New Features**
  - Added local usage reporting for Claude Code, including token totals, estimated costs, date filtering, duplicate handling, and scan diagnostics.
  - Added public APIs to discover supported local tools, create readers, register readers, and aggregate usage reports.
  - Added shared types describing usage totals, scan results, reader metadata, failures, and cost confidence.

- **Documentation**
  - Added comprehensive API reference documentation for local usage APIs and types.

- **Tests**
  - Added continuous coverage for scanning, deduplication, malformed data, missing stores, errors, and repeatability.

### CODE REVIEW

**Comment by Tara-ag** [comment]
## Review Summary for PR #1482

**Decision: APPROVED** - No blocking issues found.

### Overview
This PR adds observability tracking for Claude Code's local token usage by reading from its session transcripts. It implements one of twelve planned readers, focusing on the most common CLI (Claude Code) to provide immediate value while establishing the pattern for future readers.

### Files Changed (8 total)
- `package.json` - Added test script entry
- `src/lib/index.ts` - Added exports for local usage API
- `src/lib/localUsage/claudeCodeReader.ts` - Core implementation (NEW)
- `src/lib/localUsage/index.ts` - Aggregator function (NEW)
- `src/lib/localUsage/localUsageReaderRegistry.ts` - Registry with dynamic import factory (NEW)
- `src/lib/types/index.ts` - Added type export
- `src/lib/types/localUsage.ts` - Type definitions (NEW)
- `test/continuous-test-suite-local-usage.ts` - E2E test suite (NEW)

### Verification Checklist

✅ **Architecture & CLAUDE.md Critical Rules:**
- Rule 1 (dynamic imports): Uses dynamic import via factory function in registry ✓
- Rule 4 (CLI ≠ SDK): Local usage tracking doesn't leak CLI-specific concerns into SDK path ✓
- Rule 5 (backward compatibility): Only adds new exports, no breaking changes to existing API ✓
- Rule 7-13 (type system): All types use `type` not `interface`, located in `src/lib/types/`, unique names with `LocalUsage*` prefix, barrel-only exports ✓
- Rule 15 (E2E tests): All tests import from `dist/index.js`, driving public surface only ✓

✅ **Security:**
- No hardcoded secrets or credentials
- No injection vulnerabilities
- Reads from user's home directory (`~/.claude/projects/`) - internal stats collection only
- Handles file read errors gracefully without data loss
- No PII exposure - only token counts and costs

✅ **Type Safety:**
- All TypeScript types properly defined
- Defensive handling of optional fields
- No `any` types or unsafe casts
- Proper error handling per-file without cascading failures

✅ **Testing:**
- 8 comprehensive test cases covering:
  - SDK exports verification
  - Deduplication with message-id-keep-max strategy
  - Subagent transcript inclusion (recursive directory walk)
  - Non-message record types skipped gracefully
  - Unpriced model tracking
  - Absent CLI detection vs failure distinction
  - Real data coherence on actual machine
  - Idempotent repeated scans

✅ **Error Handling:**
- Per-file read errors reported but don't cascade
- Torn JSON lines during live sessions handled gracefully
- Time-based filtering respects mtime boundaries

### Impact Assessment
- **Blast radius**: LOW - isolated to new `localUsage` module
- **Affected flows**: None - reads from filesystem, doesn't modify existing code paths
- **API changes**: Additive only - new exports, no modifications to existing APIs
- **Dependencies**: Only uses Node.js built-ins (fs, os, path, readline)

### Minor Notes
- This is the first of twelve planned readers; future readers will follow the same pattern
- The design intentionally separates "not installed" from "failed" for better UX
- Cost confidence stays "modeled" even with unpriced models (correct - we know rates but not which turns used them)

### Conclusion
The implementation is clean, well-tested, and follows all project standards. Ready to merge.

**Review by Tara-ag — APPROVED** [review]

**Comment by Tara-ag** [comment]
## Review Summary

### Decision: APPROVED ✅

This PR adds observability functionality to read token usage from Claude Code's local session transcripts. The implementation is well-structured and follows NeuroLink's architectural patterns.

### Changes Reviewed

**New Files (8 total):**
1. `src/lib/localUsage/claudeCodeReader.ts` - Core reader implementation for Claude Code transcripts
2. `src/lib/localUsage/index.ts` - Module barrel exports  
3. `src/lib/localUsage/localUsageReaderRegistry.ts` - Factory+registry pattern
4. `src/lib/types/localUsage.ts` - Type definitions
5. `src/lib/index.ts` - Updated to export new localUsage functions
6. `src/lib/types/index.ts` - Updated to re-export localUsage types
7. `test/continuous-test-suite-local-usage.ts` - Comprehensive end-to-end tests
8. `package.json` - Added test script

### Quality Assessment

✅ **Architecture**: Follows factory+registry pattern consistent with provider system
✅ **Error Handling**: Properly handles file read errors without breaking scans
✅ **Memory Efficiency**: Uses line-by-line reading for large files (tens of MB)
✅ **Type Safety**: All types properly defined in canonical location (`src/lib/types/`)
✅ **Testing**: 9 comprehensive E2E test cases covering edge cases
✅ **Backward Compatibility**: No breaking changes, pure additive feature
✅ **Documentation**: Well-commented code with clear rationale

### Test Coverage

The test suite covers:
- SDK exports verification
- Deduplication strategy (keep-max per message ID)
- Subagent transcript inclusion
- Non-message record handling
- Unpriced model reporting
- Missing store detection
- Real data validation at scale (17k+ files, ~10 GB)
- Idempotency across repeated scans

### Impact Analysis

- **Blast Radius**: Low - only adds new API surface, no modifications to existing providers
- **Affected Flows**: None - self-contained feature
- **Dependencies**: None - reads filesystem directly, no external dependencies

### Recommendations

None - this is production-ready code that follows all project standards and conventions.

---

**Review Scope**: Live review of PR #1482 "feat(localUsage): read Claude Code token spend from its own transcripts"

**Review by Tara-ag — APPROVED** [review]

**Review by Tara-ag — COMMENTED** [review]
## Review Summary

This PR adds local usage tracking for Claude Code, reading token spend from its own session transcripts. This is a well-architected feature that follows the existing factory+registry pattern used throughout NeuroLink.

### Key Observations

✅ **Architecture**: Follows CLAUDE.md Rule 1 - uses dynamic imports in registry factory function  
✅ **Types**: All type definitions in canonical `src/lib/types/` location (Rule 2)  
✅ **Type aliases**: Uses `type`, never `interface` (Rule 7)  
✅ **Barrel imports**: Uses `export *` only (Rule 10)  
✅ **Error handling**: Returns errors, never throws (Rule 6)  
✅ **Testing**: End-to-end test suite imports from dist, not src (Rule 15)  
✅ **Documentation**: API reference updated with new exports  

### Findings

- **MINOR**: Test coverage gap for collectTranscripts - no direct tests for edge cases like unreadable directories or permission errors. Add a test case for when readdir throws due to permissions, and verify it returns silently without aborting the scan.

### Impact Analysis

- Self-contained feature with minimal cross-file impact (~118 files affected through documentation updates)
- No breaking changes to existing API
- No security concerns (filesystem-based, no credentials involved)
- Proper error isolation - individual reader failures don't cascade

### Decision

**APPROVED** - The implementation is sound and follows all architectural guidelines. One minor test coverage improvement recommended but not blocking.

File: src/lib/localUsage/claudeCodeReader.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+ * Read line by line rather than into a string: these files reach tens of
+ * megabytes, and a full-history scan opens thousands of them.
+ *
+ * Dedup is by `message.id`, keeping the LARGEST output count seen for that id.
+ * A resumed session re-logs turns it has already written, and the re-logged
+ * copy can carry a higher output count than the first — taking the max rather
+ * than the first or last is what makes a resumed session total correct instead
+ * of either under- or double-counted. The map is per-file and discarded after,
+ * so it stays bounded no matter how many files a scan covers.
+ */
```
Comment (Tara-ag) [inline src/lib/localUsage/claudeCodeReader.ts:103]: 💡 MINOR: Test coverage gap for collectTranscripts - The collectTranscripts function is tested indirectly through foldTranscript but has no direct tests for edge cases like unreadable directories or permission errors. Add a test case for when readdir throws due to permissions, and verify it returns silently without aborting the scan.

**Review by Tara-ag — APPROVED** [review]
## Review Summary

**Decision: APPROVED** ✅

This PR adds observability functionality to read token usage from Claude Code's local session transcripts. The implementation is well-structured and follows NeuroLink's architectural patterns.

### Changes Reviewed

**New Files (8 total):**
1. `src/lib/localUsage/claudeCodeReader.ts` - Core reader implementation for Claude Code transcripts
2. `src/lib/localUsage/index.ts` - Module barrel exports
3. `src/lib/localUsage/localUsageReaderRegistry.ts` - Factory+registry pattern
4. `src/lib/types/localUsage.ts` - Type definitions
5. `src/lib/index.ts` - Updated to export new localUsage functions
6. `src/lib/types/index.ts` - Updated to re-export localUsage types
7. `test/continuous-test-suite-local-usage.ts` - Comprehensive end-to-end tests
8. `package.json` - Added test script

### Quality Assessment

✅ **Architecture**: Follows factory+registry pattern consistent with provider system, uses dynamic imports in registry (CLAUDE.md Rule 1)
✅ **Error Handling**: Properly handles file read errors without breaking scans
✅ **Memory Efficiency**: Uses line-by-line reading for large files (tens of MB)
✅ **Type Safety**: All types properly defined in canonical location (`src/lib/types/`)
✅ **Testing**: 9 comprehensive E2E test cases covering edge cases
✅ **Backward Compatibility**: No breaking changes, pure additive feature
✅ **Documentation**: Well-commented code with clear rationale

### Findings

- **MINOR**: Test coverage gap for collectTranscripts - no direct tests for edge cases like unreadable directories or permission errors (inline comment added above)

### Impact Analysis

- **Blast Radius**: Low - only adds new API surface, no modifications to existing providers
- **Affected Flows**: None - self-contained feature
- **Dependencies**: None - reads filesystem directly, no external dependencies

### Conclusion

This is production-ready code that follows all project standards and conventions. Ready to merge.

**Review by Tara-ag — APPROVED** [review]
## Review Summary

**Decision: APPROVED** ✅

This PR adds observability functionality to read token usage from Claude Code's local session transcripts. The implementation is well-structured and follows NeuroLink's architectural patterns.

### Changes Reviewed

**New Files (8 total):**
1. `src/lib/localUsage/claudeCodeReader.ts` - Core reader implementation for Claude Code transcripts
2. `src/lib/localUsage/index.ts` - Module barrel exports
3. `src/lib/localUsage/localUsageReaderRegistry.ts` - Factory+registry pattern
4. `src/lib/types/localUsage.ts` - Type definitions
5. `src/lib/index.ts` - Updated to export new localUsage functions
6. `src/lib/types/index.ts` - Updated to re-export localUsage types
7. `test/continuous-test-suite-local-usage.ts` - Comprehensive end-to-end tests
8. `package.json` - Added test script

### Quality Assessment

✅ **Architecture**: Follows factory+registry pattern consistent with provider system, uses dynamic imports in registry (CLAUDE.md Rule 1)
✅ **Error Handling**: Properly handles file read errors without breaking scans
✅ **Memory Efficiency**: Uses line-by-line reading for large files (tens of MB)
✅ **Type Safety**: All types properly defined in canonical location (`src/lib/types/`)
✅ **Testing**: 9 comprehensive E2E test cases covering edge cases
✅ **Backward Compatibility**: No breaking changes, pure additive feature
✅ **Documentation**: Well-commented code with clear rationale

### Findings

- **MINOR**: Test coverage gap for collectTranscripts - no direct tests for edge cases like unreadable directories or permission errors (inline comment added above)

### Impact Analysis

- **Blast Radius**: Low - only adds new API surface, no modifications to existing providers
- **Affected Flows**: None - self-contained feature
- **Dependencies**: None - reads filesystem directly, no external dependencies

### Conclusion

This is production-ready code that follows all project standards and conventions. Ready to merge.

**Comment by Tara-ag** [comment]
## 🛡️ Yama Review Verdict: CHANGES_REQUESTED

Severity counts — 🔒 CRITICAL: 0 · ⚠️ MAJOR: 0 · 💡 MINOR: 1

## 🤖 Yama Review Summary

⚠️ The review loop ended early (`time-limit`, 28 steps) — this verdict is built strictly from gate-verified findings.

**Verified findings (1):**

- **MINOR**: Test coverage gap for collectTranscripts — `src/lib/localUsage/claudeCodeReader.ts:103`

### Findings behind this verdict
- 💡 **MINOR**: Test coverage gap for collectTranscripts — `src/lib/localUsage/claudeCodeReader.ts:103`
  The collectTranscripts function is tested indirectly through foldTranscript but has no direct tests for edge cases like unreadable directories or permission errors. Add a test case for when readdir throws due to permissions, and verify it returns silently without aborting the scan.

File: src/lib/localUsage/claudeCodeReader.ts (hunk 1 above)
Comment (murdore) [reply src/lib/localUsage/claudeCodeReader.ts:103]: Added. `test/continuous-test-suite-local-usage.ts` now has "an unreadable project directory is skipped, not fatal": it writes a readable transcript plus a sibling project directory chmod'd to 0o000 containing a transcript that must not be counted, then asserts the scan completes, reports no reader failure, and returns exactly the readable file's totals.

Two things it guards against beyond the obvious. It first checks that this process genuinely cannot read a 0o000 directory and SKIPs if it can — running as root would defeat the permission bit and the case would pass while proving nothing. And it restores the mode inside the scan callback rather than in teardown, because a 0o000 directory defeats `rmSync` too; the first version failed with ENOTEMPTY from its own cleanup, which read as a failure of the code under test.

Confirmed non-vacuous by making `collectTranscripts` rethrow instead of returning: the case fails with "an unreadable directory aborted the whole scan". Suite is 9 passing.

**Comment by Tara-ag** [comment]
🔍 MINOR: The num() function at line 89 lacks documentation explaining its purpose and behavior when parsing invalid numbers. This helper function is used throughout the file to parse cost strings, and adding JSDoc would help future maintainers understand why it returns 0 for invalid inputs instead of throwing an error.

**Comment by Tara-ag** [comment]
🔍 MINOR: The finally block at line 137 calls fs.chmodSync to restore permissions, but doesn't handle the case where the locked directory might have been deleted before reaching this point. This could cause an unnecessary error when the scan completes normally. Add a check to verify the locked directory still exists before attempting to restore permissions.

File: docs/api/type-aliases/LocalUsageAggregateReport.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+---
+
+### failures
+
+> **failures**: [`LocalUsageReaderFailure`](LocalUsageReaderFailure.md)[]
+
+Defined in: [types/localUsage.ts:154](https://github.com/juspay/neurolink/blob/release/src/lib/types/localUsage.ts#L154)
+
+CLIs that were registered but produced nothing, and why.
+
```
[the remark this answers is not in the document]
Comment (murdore) [reply docs/api/type-aliases/LocalUsageAggregateReport.md:42]: Corrected in #1496. The doc comment now reads "CLIs whose reader could not be created, detected, or scanned, and why." You were right that the old wording was actively misleading rather than just vague: an absent store is reported under notInstalled, and calling that "produced nothing" made the two indistinguishable to a caller trying to tell a broken reader from an uninstalled CLI.

File: src/lib/localUsage/claudeCodeReader.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+      const unpriced = new Set<string>();
+
+      const files: string[] = [];
+      await collectTranscripts(projectsRoot(), files);
+
+      const sinceDays = options?.sinceDays ?? DEFAULT_SINCE_DAYS;
+      const cutoff =
+        Number.isFinite(sinceDays) && sinceDays > 0
+          ? Date.now() - sinceDays * 86_400_000
+          : undefined;
```
[the remark this answers is not in the document]
Comment (murdore) [reply src/lib/localUsage/claudeCodeReader.ts:226]: Real bug, fixed in #1496 — and worse than the finding states. Measured on release before the fix: sinceDays 0 scanned 17,534 files in 35.9s. Infinity is now the only full-sweep signal and the window is clamped at zero, applied to all three readers since they shared the expression. After: 0 files in 1.9s for 0, same for -5, and Infinity still reads all 17,537. The CLI is unaffected — `--since 0` still means all history because the command translates it to Infinity before it reaches a reader, so that convention no longer depends on the two agreeing by accident. Confirmed non-vacuous by restoring the old guard: the new case fails with "sinceDays 0 read transcripts — got 1 turns instead of 0".
